src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs

```diff
@@ -1271,21 +1269,30 @@ public override SortedSet<string> Terms(string field, string fromValue, long pag
                 mltQuery = IndexSearcher.And(mltQuery, moreLikeThisQuery.FilterQuery);
             }
 
-
-
             var ravenIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
             long[] ids = QueryPool.Rent(pageSize);
             var read = 0;
+            long returnedDocs = 0;
+            long skippedDocs = 0;
             Page page = default;
             while ((read = mltQuery.Fill(ids.AsSpan())) != 0)
             {
                 for (int i = 0; i < read; i++)
                 {
+                    if (returnedDocs >= query.Limit)
+                        yield break;
+                    
                     var hit = ids[i];
                     token.ThrowIfCancellationRequested();
 
                     if (hit == baseDocId)
                         continue;
+                    
+                    if (skippedDocs < query.Start)
+                    {
+                        skippedDocs++;
+                        continue;
+                    }
```

Comment:
You've to to this after registering id in `ravenIds` since it could lead to invalid paging when dealing with fanout index.
Situation like
`[doc0, ..., doc9, doc9, doc10]` (only doc9 has two index entries).
So, if you take all, the second doc9 will be discarded, however when you do pageSize 10:
`[doc0, ..., doc9], [doc9, doc10]`
but doc9 should not be in second batch.

Reply:
Moved